### PR TITLE
Issue #15943: Fix try with resources AST

### LINKED ISSUE #15943: ArrayIndexOutOfBoundsException on indentation check of try-with-resources
I have read check documentation: https://checkstyle.sourceforge.io/checks/misc/indentation.html#Indentation
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac Thingie.java
no output (success)

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <property name="severity" value="error"/>

    <property name="fileExtensions" value="java, properties, xml"/>

    <module name="TreeWalker">
        <module name="Indentation">
        </module>
    </module>
</module>

/var/tmp $ cat Thingie.java
public class Thingie implements AutoCloseable {
    public void doIt() {
        open();
        try (this) {
            System.out.println("doIt");
        }
    }
    
    public void open() {
        System.out.println("open");
    }
    
    @Override
    public void close() {
        System.out.println("close");
    }
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar  checkstyle-10.20.1-all.jar -c config.xml Thingie.java
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing Thingie.java
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:312)
	at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:226)
	at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:415)
	at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:338)
	at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:195)
	at com.puppycrawl.tools.checkstyle.Main.main(Main.java:130)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 2147483647 out of bounds for length 17
	at com.puppycrawl.tools.checkstyle.api.FileText.get(FileText.java:222)
	at com.puppycrawl.tools.checkstyle.api.FileContents.getLine(FileContents.java:100)
	at com.puppycrawl.tools.checkstyle.api.AbstractCheck.getLine(AbstractCheck.java:308)
	at com.puppycrawl.tools.checkstyle.checks.indentation.AbstractExpressionHandler.getLineStart(AbstractExpressionHandler.java:224)
	at com.puppycrawl.tools.checkstyle.checks.indentation.AbstractExpressionHandler.getLineStart(AbstractExpressionHandler.java:213)
	at com.puppycrawl.tools.checkstyle.checks.indentation.AbstractExpressionHandler.isOnStartOfLine(AbstractExpressionHandler.java:178)
	at com.puppycrawl.tools.checkstyle.checks.indentation.TryHandler.checkTryResources(TryHandler.java:150)
	at com.puppycrawl.tools.checkstyle.checks.indentation.TryHandler.checkIndentation(TryHandler.java:83)
	at com.puppycrawl.tools.checkstyle.checks.indentation.IndentationCheck.visitToken(IndentationCheck.java:435)
	at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:393)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:466)
	at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:331)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:211)
	at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:101)
	at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:340)
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:299)
	... 5 more
Checkstyle ends with 1 errors.
```
---

I expect no ArrayIndexOutOfBoundsException exception to be thrown.

```
$ java -jar  checkstyle-10.20.1-all.jar -t Test.java
COMPILATION_UNIT -> COMPILATION_UNIT [1:0]
`--CLASS_DEF -> CLASS_DEF [1:0]
    |--MODIFIERS -> MODIFIERS [1:0]
    |   `--LITERAL_PUBLIC -> public [1:0]
    |--LITERAL_CLASS -> class [1:7]
    |--IDENT -> Thingie [1:13]
    |--IMPLEMENTS_CLAUSE -> implements [1:21]
    |   `--IDENT -> AutoCloseable [1:32]
    `--OBJBLOCK -> OBJBLOCK [1:46]
        |--LCURLY -> { [1:46]
        |--METHOD_DEF -> METHOD_DEF [2:4]
        |   |--MODIFIERS -> MODIFIERS [2:4]
        |   |   `--LITERAL_PUBLIC -> public [2:4]
        |   |--TYPE -> TYPE [2:11]
        |   |   `--LITERAL_VOID -> void [2:11]
        |   |--IDENT -> doIt [2:16]
        |   |--LPAREN -> ( [2:20]
        |   |--PARAMETERS -> PARAMETERS [2:21]
        |   |--RPAREN -> ) [2:21]
        |   `--SLIST -> { [2:23]
        |       |--EXPR -> EXPR [3:12]
        |       |   `--METHOD_CALL -> ( [3:12]
        |       |       |--IDENT -> open [3:8]
        |       |       |--ELIST -> ELIST [3:13]
        |       |       `--RPAREN -> ) [3:13]
        |       |--SEMI -> ; [3:14]
        |       |--LITERAL_TRY -> try [4:8]
        |       |   |--RESOURCE_SPECIFICATION -> RESOURCE_SPECIFICATION [4:12]
        |       |   |   |--LPAREN -> ( [4:12]
        |       |   |   |--RESOURCES -> RESOURCES [-2147483648:-2147483648]
        |       |   |   |   `--RESOURCE -> RESOURCE [-2147483648:-2147483648]
        |       |   |   `--RPAREN -> ) [4:17]
        |       |   `--SLIST -> { [4:19]
        |       |       |--EXPR -> EXPR [5:30]
        |       |       |   `--METHOD_CALL -> ( [5:30]
        |       |       |       |--DOT -> . [5:22]
        |       |       |       |   |--DOT -> . [5:18]
        |       |       |       |   |   |--IDENT -> System [5:12]
        |       |       |       |   |   `--IDENT -> out [5:19]
        |       |       |       |   `--IDENT -> println [5:23]
        |       |       |       |--ELIST -> ELIST [5:31]
        |       |       |       |   `--EXPR -> EXPR [5:31]
        |       |       |       |       `--STRING_LITERAL -> "doIt" [5:31]
        |       |       |       `--RPAREN -> ) [5:37]
        |       |       |--SEMI -> ; [5:38]
        |       |       `--RCURLY -> } [6:8]
        |       `--RCURLY -> } [7:4]
...

```

### PR DESCRIPTION
**Pull request by nrmancuso**
Closes #15943

Report generation: https://github.com/nrmancuso/checkstyle-diff-report-generator/actions/runs/11967064218

—————-

Links:
Projects file: projects-to-test-on.properties
Patch branch: issue-15934

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/issue-15934/2024-11-22-T-07-32-55/part1/index.html

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/issue-15934/2024-11-22-T-07-32-55/sevntu-check-regression_part_2/index.html

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/issue-15934/2024-11-22-T-07-32-55/part6/index.html

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/issue-15934/2024-11-22-T-07-32-55/part5/index.html

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/issue-15934/2024-11-22-T-07-32-55/part4/index.html

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/issue-15934/2024-11-22-T-07-32-55/sevntu-check-regression_part_1/index.html

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/issue-15934/2024-11-22-T-07-32-55/part2/index.html

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/issue-15934/2024-11-22-T-07-32-55/part3/index.html

All diffs are in noncompilable files, in the stack trace.

https://checkstyle-reports.s3.us-east-2.amazonaws.com/reports/issue-15934/2024-11-22-T-07-32-55/antlr-report-checkstyle/index.html

diff in ANTLR report is also from changes in exceptions in files that typically cause stackoverflow when the AST is printed.

### CODE REVIEW

File: src/test/resources/com/puppycrawl/tools/checkstyle/grammar/ExpectedAstRegressionTryWithResourcesOnAutoCloseable.txt [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        |       |       |       |--IDENT -> open [6:12]
+        |       |       |       |--ELIST -> ELIST [6:17]
+        |       |       |       `--RPAREN -> ) [6:17]
+        |       |       |--SEMI -> ; [6:18]
+        |       |       |--LITERAL_TRY -> try [7:12]
+        |       |       |   |--RESOURCE_SPECIFICATION -> RESOURCE_SPECIFICATION [7:16]
+        |       |       |   |   |--LPAREN -> ( [7:16]
+        |       |       |   |   |--RESOURCES -> RESOURCES [7:17]
+        |       |       |   |   |   `--RESOURCE -> RESOURCE [7:17]
+        |       |       |   |   |       `--LITERAL_THIS -> this [7:17]
```
Comment (nrmancuso) [inline src/test/resources/com/puppycrawl/tools/checkstyle/grammar/ExpectedAstRegressionTryWithResourcesOnAutoCloseable.txt:51]: The change in AST is here:
```suggestion
        |       |       |   |   |--RESOURCES -> RESOURCES [-2147483648:-2147483648]
        |       |       |   |   |   `--RESOURCE -> RESOURCE [-2147483648:-2147483648]
```

File: src/test/resources/com/puppycrawl/tools/checkstyle/grammar/InputAstRegressionTryWithResourcesOnAutoCloseable.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+            open();
+            try (this) {
+                System.out.println("doIt");
+            }
+
+            T t = new T();
+
+            try (t) {
+                System.out.println("doIt");
+            }
```
Comment (nrmancuso) [inline src/test/resources/com/puppycrawl/tools/checkstyle/grammar/InputAstRegressionTryWithResourcesOnAutoCloseable.java:15]: Made sure we had coverage for the single `ident` as well.

**Review by nrmancuso — COMMENTED** [review]
Comments:

**Comment by nrmancuso** [comment]
We can proceed with review while reports are being generated

**Review by mahfouz72 — APPROVED** [review]
LGTM. Thanks a lot.

**Comment by romani** [comment]
I don't see some explanation on why AST should be changed, this is unusual fix  ... By fix of AST.

**Comment by mahfouz72** [comment]
I dont see this issue as an indentation check bug. It is AST bug indeed, that we discovered through the check execution. I don't get why this is an unusual fix?

**Review by romani — APPROVED** [review]
ok to merge.

thanks a lot.

**Pull request merged**